### PR TITLE
ZTS: add `snapdir` test group

### PR DESCRIPTION
**Pull request by robn**
_[Sponsors: TrueNAS]_

### Motivation and Context

We're not really testing the behaviour of the snapshot mounts directly, beyond actually using them and a couple of basic tests in the `snapshot` group.

While rewriting that entire subsystem for Linux (coming soon!), I found a number of very subtle edge cases that we weren't considering at all. I built tests for them as I went, and here they are.

Note that FreeBSD's implementation of this feature is simpler and has no tunables, so unless otherwise noted, these tests only apply to Linux.

### Description

This creates a `snapdir` test group, specifically focused on behaviours and semantics of operations under `.zfs/snapshot`.

The three tests `snapshot_015_pos`, `snapshot_016_pos` and `snapshot_019_pos` all dealt with snapdir specifics, and so are moved and renamed `snapdir_admin_create_destroy`, `snapdir_admin_rename_destroy` and `snapdir_mount_parallel` (`snapdir_mount_parallel` applies to all platforms).

Existing `snapshot` tests are adjusted to not attempt to remove the snapdir during cleanup, which they don't actually need to do, and so we can remove the need to touch the `ADMIN_SNAPSHOT` (`zfs_admin_snapshot`) tunable in that group.

Then there's new tests:
- `snapdir_admin_disabled` ensures that the when `zfs_admin_snapshot=0`, the "magic" to create/destroy/rename a snapshot by doing that to the snapshot dir is not allowed
- `snapdir_mount_unmount` checks that if automounted snapshot is manually unmount by the admin, re-entering the snapdir will cause a new mount.
- `snapdir_mount_expire_idle` and `snapdir_mount_expire_busy` check that an idle snapshot mount is eventually "expired" and unmounted, but a busy mount will not be expired. The `zfs_expire_snapshot` tunable is used to control the expiry timeout.
- `snapdir_mount_no_setuid` ensures that snapshout mounts have option `suid` or `nosuid` depending on the how the `zfs_snapshot_no_setuid` tunable is set.
- `snapdir_stat_no_mount` makes sure that ops on the snapshot dir itself do not trigger a mount; rather, the mount is only triggered by accesses "beyond" the snapshot dir. This is somewhat redundant, confirming the standard Linux automount behaviour, however, in the new implementation I have to support this explicitly (for reasons which will explained there), so confirming that behaviour in test seems harmless at best. This test is Linux specific as the setup is different on FreeBSD - the mount occurs on first access there.

~The top commit adds four tests, 'snapdir_mount_bind', 'snapdir_mount_move', 'snapdir_mount_expire_bind', and `snapdir_mount_expire_move`. These all test that if the automount is moved away (`mount --move`) or cloned (`mount --bind`) and the original unmounted, then a new automount can be triggered on the now-empty snapdir, and the "other" mount will not be expired. Currently these tests fail because of an architectural issue in the implementation, which is part of why I've been rewriting it. So I have added these tests to the "know failure" list at the same time, but I would be happy to just drop this commit and include it in that PR if preferred.~

Update: [dropped bind and move tests](https://github.com/openzfs/zfs/pull/18705#issuecomment-4840995545).

### How Has This Been Tested?

The `snapshot` and `snapdir` test groups have been run on Linux 7.0.8, 6.12.84 and 5.10.253, with same results on all:

- `snapdir`: PASS 13
- `snapshot`: PASS 22

Also run on FreeBSD 15.1-RELEASE:

- `snapdir`: PASS 4
- `snapshot`: PASS 22

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by robn** [comment]
Four of the five CI failures look like misc CI timeouts/glitches, not worrying about that yet.

The fifth is curious - it looks like a true failure, as in, the mount happened when it shouldn't have. It certainly doesn't on the same kernel series (7.0) here. So I've thrown in some extra checks in that test just for the CI run. One to make sure its not already mounted before the non-trigger under test, and another to wait a second between the trigger and the test. Neither would tell me much conclusively, just suggest another direction to look.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_admin_rename_destroy.ksh [hunk 1]
```diff
@@ -57,11 +61,15 @@ function cleanup
...
 	log_must zfs create $TESTPOOL/$TESTFS
 	log_must zfs set mountpoint=$TESTDIR $TESTPOOL/$TESTFS
+
+	restore_tunable ADMIN_SNAPSHOT
 }
 
 log_assert "Verify renamed snapshots via mv can be destroyed."
 log_onexit cleanup
 
+log_must set_tunable64 ADMIN_SNAPSHOT 1
```
Comment (ixhamza) [inline tests/zfs-tests/tests/functional/snapdir/snapdir_admin_rename_destroy.ksh:73]: Should we `save_tunable` `ADMIN_SNAPSHOT` first? otherwise cleanup's `restore_tunable` no-ops.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_mount_expire_move.ksh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+log_must test "$(get_mount_paths $TESTPOOL/$TESTFS@snap)" == "$TESTDIR/fs/$SNAPROOT/snap"
+
+# Move the mount away
+log_must mkdir $TESTDIR/move1
+log_must mount --move $TESTDIR/fs/$SNAPROOT/snap $TESTDIR/move1
+log_must test "$(get_mount_paths $TESTPOOL/$TESTFS@snap)" == "$TESTDIR/move1"
+
+# Wait for the expiry, then ensure the moved mount remains
+log_note "sleeping until snapshot expiry time has passed"
+sleep 4
```
Comment (ixhamza) [inline tests/zfs-tests/tests/functional/snapdir/snapdir_mount_expire_move.ksh]: Default expire is 300s, so this never actually expires.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_mount_no_setuid.ksh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+log_must zfs snapshot $TESTPOOL/$TESTFS@snap1
+log_must zfs snapshot $TESTPOOL/$TESTFS@snap2
+
+# Set the tunable to enable setuid.
+log_must set_tunable64 SNAPSHOT_NO_SETUID 0
+
+# Trigger automount on the first snapshot, verify it does not have the
+# 'nosuid' option.
+ls $TESTDIR/$SNAPROOT/snap1
+mount | grep $TESTPOOL/$TESTFS@snap1 | grep -q nosuid
```
Comment (ixhamza) [inline tests/zfs-tests/tests/functional/snapdir/snapdir_mount_no_setuid.ksh]: Should we confirm `snap1` mounted first? this passes even if it didn't.

**Review by ixhamza — COMMENTED** [review]
Is [XXX snapdir_stat_no_mount: CI hacks](https://github.com/openzfs/zfs/pull/18705/changes/03308e2df7b5b3aa8a1714986c1a04b1318df41f) still needed?

**Comment by robn** [comment]
Figured out what's up with `snapdir_stat_no_mount`. It's actually a coreutils vs uutils difference. coreutils calls `statx()` with `AT_NO_AUTOMOUNT` when doing an existence check, uutils does not. The thing I'm actually wanting to test is that `AT_NO_AUTOMOUNT` is honoured, so I'll extend one of our helpers instead.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_admin_rename_destroy.ksh (hunk 1 above)
Comment (robn) [reply tests/zfs-tests/tests/functional/snapdir/snapdir_admin_rename_destroy.ksh:73]: Ahh yep, just missed when removing the "global" tunable set. Fixed.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_mount_expire_move.ksh (hunk 2 above)
Comment (robn) [reply tests/zfs-tests/tests/functional/snapdir/snapdir_mount_expire_move.ksh]: Bummer, missed it. Danger of "fail by design" tests. Fixed, and confirmed passing ie expiring correctly on the rewrite branch.

File: tests/zfs-tests/tests/functional/snapdir/snapdir_mount_no_setuid.ksh (hunk 3 above)
Comment (robn) [reply tests/zfs-tests/tests/functional/snapdir/snapdir_mount_no_setuid.ksh]: Great catch. Replaced with a string substitution and before/after comparison. Much more solid.

**Comment by robn** [comment]
> Is [XXX snapdir_stat_no_mount: CI hacks](https://github.com/openzfs/zfs/pull/18705/changes/03308e2df7b5b3aa8a1714986c1a04b1318df41f) still needed?

Removed, now I know what's happening.

`snapdir_stat_no_mount` is now using our `statx` test helper, not for any particular thing it gives, but just because it calls `open(..., O_PATH)` only, which does not trigger automount.

The actual thing seems to be that most syscalls just do the thing they're asked, and automount may or may not happen depending on the nature of the pathwalk. However, `statx()` _will_ trigger automount if `AT_NO_AUTOMOUNT` is not set.

The issue for Ubuntu 25+ is that their `ls(1)` comes from uutils, a reimplementation of coreutils. It appears that that calls into the Rust stdlib for `stat()`, and that uses `statx()` without `AT_NO_AUTOMOUNT`. So its not just about `ls`, even using `/usr/bin/stat` yields the same result.

Further reading:
- https://bugs.launchpad.net/ubuntu/+source/rust-coreutils/+bug/2130843
- https://github.com/uutils/coreutils/issues/9478

While our `statx` helper _is_ calling `statx()` without `AT_NO_AUTOMOUNT`, that's _after_ it does the path resolution with a plain old `open()`, so no changes are required there.

**Comment by robn** [comment]
I've dropped the bind and move tests from this PR for now. I thought it might be useful to have failing tests that were obviously the "wrong" behaviour, but today I realised that bind and move are far more complicated than I realised. I know we have the wrong behaviour right now, but I'll need to think harder about what the "right" behaviour is, so I'll include those tests in a future PR that implements the right thing.

**Review by ixhamza — APPROVED** [review]

**Review by behlendorf — APPROVED** [review]
